kvserver: remove ApplyImpact

wenyihu6 opened the pull request:
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151231)
<!-- Reviewable:end -->

wenyihu6 committed:
mmaintegration: add  allocator_op.go and allocator_sync.go

This commit adds allocator_op.go and allocator_sync.go, introducing
trackedAllocatorChange and basic coordination helpers for allocator sync. These
helpers register changes before they are applied and post apply the effect on
store pool after the change is applied. Note that allocator sync is not plumbed
or used yet. Future commits would plumb it to replicate queue, lease queue, and
mma store rebalancer to coordinate changes.

wenyihu6 committed:
kvserver: register external changes with mma

Previously, allocator sync was called before and after applying changes, but
only to update the store pool. This commit extends it to also register changes
with the mma allocator. If mma is enabled at registration time, allocator sync
tracks the change and informs mma during PostApply. If mma is disabled, it
continues to only update the store pool.

wenyihu6 committed:
mmaintegration: add store load msg

This commit adds MakeStoreLoadMsg, moved from allocator_mma_integration.go in
mmaprototypehelpers. Eventually, all functions from mmaprototypehelpers will be
migrated to pkg/kv/kvserver/mmaintegration as part of productionizing the code.
Note that no tests have been added yet - a TODO has been left in place to
address this. Added to the spreadsheet as well.

wenyihu6 committed:
mmaintegration: remove pkg/kv/kvserver/allocator/mmaprototypehelpers

Previously, all helper functions from
pkg/kv/kvserver/allocator/mmaprototypehelpers were migrated to
pkg/kv/kvserver/mmaintegration. This commit completes the transition by removing
all usages of mmaprototypehelpers to pkg/kv/kvserver/mmaintegration.

wenyihu6 committed:
mmaintegration: add InvalidSyncChangeID

This commit adds InvalidSyncChangeID to allocator sync, used only by asim to
indicate that no change is currently in progress. This is needed because asim
returns control to the tick loop rather than waiting for changes to complete
during simulations. So it needs to track the current change id in progress.

wenyihu6 committed:
mmaintegration: integrate allocator sync to mma store rebalancer

Previously, mma PreApply was introduced to register mma changes with allocator
sync. This commit integrates it into the mma store rebalancer. To handle
failures that occur before a change is even registered (such as when a replica
doesn’t exist), this commit also adds, MarkChangesAsFailed, in allocator sync.
This function bypasses the trackedChanges map since the change is not meant to
be applied. It must only be called on failure, and allocator sync is responsible
for notifying MMA of the failure.

wenyihu6 committed:
kvserver: use allocator sync to update store pool

Previously, allocator sync was plumbed into the replicate queue, lease queue,
and mma store rebalancer. This commit updates the lease and replicate queues to
call allocator sync to register changes before they are applied, and again post
apply to reflect load changes in the local store pool. For now, allocator sync
is only used for updating the store pool. Future commits will extend this to
coordinate with mma.

wenyihu6 committed:
kvserver: plumb allocator sync

Previously, allocator sync files were added. This commit wires allocator sync
into the lease queue, replicate queue, and mma store rebalancer. Each node has a
single allocator sync instance responsible for coordinating with the
corresponding mma allocator. Note that while the allocator_sync is now plumbed,
they are not yet actively used.

wenyihu6 committed:
mmaintegration: add MMAPreApply to allocator sync

Previously, allocator sync tracked external changes from the replicate and lease
queues. This commit extends it to register mma driven changes as well. On
PostApply, allocator sync now informs mma of the outcome. Note that it is not
being used yet. Future commits will call into it from mma store rebalancer.

wenyihu6 committed:
mmaintegration: add interface for store pool and mma

This commit integrates the store pool and MMA allocator into allocator sync via
an interface to facilitate testing within this package.

wenyihu6 committed:
mmaintegration: address minor code review comments

This commit address minor code review comments.

wenyihu6 committed:
kvserver: update allocator sync on errors in AdminTransferLease

Previously, the replicate queue could skip updating allocator sync when
rlm.AdminTransferLease failed, due to early returns on error. This commit
ensures allocator sync is always updated before exiting the function, regardless
of success or failure.

wenyihu6 committed:
mmaprototype: add MakePendingRangeChangeForTesting

This commit improves the String methods for several types in mmaprototype to
make datadriven test output more readable. It also exports a
MakePendingRangeChangeForTesting helper to facilitate constructing
pendingReplicaChanges in future datadriven tests.

wenyihu6 committed:
mmaintegration: add mmaintegration datadriven tests

This commit adds a datadriven test for the mmaintegration package, verifying
the basic functionality of allocator sync. It tests how replica changes are
converted to mma replica changes and how the store pool is checked, using mocks
for components like the store pool and mma allocator.

wenyihu6 committed:
mmaintegration: add store load datadriven tests

This commit adds a datadriven test for MakeStoreLoadMsg, covering a basic store
load scenario and one with near-zero cpu and byte size utilization.
